Fix csvdiff script

milljm opened the pull request:
Fixed a few discovered issues with csvdiff.py

Removed old no-longer-used csvdiffer unittest and added a new set of unittests, specific to our stand-alone csvdiff.py script.

Added documentation.

Removed the deprecation message when using override tolerances functionality, as we want to continue to support it within test spec files.

Moved the csvdiff.py script, to moose/scripts. As it is a 'more' friendly location users are already familiar with.

Small un-related fix to the TestHarness documentation for the justification formatting of some of the tables.

Commit:
Fix parsing of the CSV comparison summary file

Add logical not capability.
Fix comparison globals not being used.
Fix comparison individual fields overriding other individual fields.
Remove the deprecation message when using override params within spec file.
Remove the old CSVDiffer unittest python file which is no longer used.
Add new CSVDiff unittests.

Closes #12687

Commit:
Adding Documentation

Adding the documentation and moving the csvdiff.py tool to a location more
easily identified as 'scripts' users know about.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/276765/) on 7ec23c6 wanted to post the following:

View the site [here](http://mooseframework.inl.gov/docs/PRs/12710/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Private App tests](https://civet.inl.gov/job/276767/) on 7ec23c6 : invalidated by @permcody

Sabertooth fixed!

moosebuild commented:
Job [Private App tests PETSc alt](https://civet.inl.gov/job/276770/) on 7ec23c6 : invalidated by @permcody

Sabertooth fixed!

Commit:
Fix a glaring issue

Add instructions on how to use the comparison file we are instructing the user to
generate.

milljm commented:
I am still finding small bugs with this. Please do not merge.

Commit:
Fields without min/max were ignored

Fix issue with comparison file parsing when a field had no commented section, it was
being ignored.

Remove all those pesky for loops!

permcody commented on scripts/csvdiff.py:
```diff
@@ -84,41 +84,39 @@ def parseComparisonFile(self, config_file):
+                    custom_params['FIELDS'][field_key[0]]['ZERO'] = self.getParamValues(words.intersection(zero_params).pop(), a_line)[0]
+
+                # Possible field containing tollerance params
+                if field_key and words.intersection(tollerance_params):
```
tollerance -> tolerance

permcody commented on modules/doc/content/python/CSVDiff.md:
```diff
@@ -0,0 +1,192 @@
+TIME STEPS relative 1 floor 0  # min: 0 @ t0  max: 1 @ t1
+
+GLOBAL VARIABLES relative 5.5e-06 floor 1e-11
+    !y                   # min: 0.000e+00 @ t0          max: 1.000e-06 @ t1
```
Wow - is this something that's supported by Exodiff?

milljm replied:
Hmmm I am not sure. I'll test it. If it does not work, I suppose I should remove this feature from csvdiff?

Or, just update our statement about both summery files being interchangeable...

permcody replied:
What did you find out about this syntax? Does this exist in Exodiff? @bwspenc - do you know?

milljm replied:
I think it does work. Here is my custom comparison file:
```
[#]> cat simple.cmp
COORDINATES absolute 1.e-6    # min separation not calculated
TIME STEPS relative 1.e-6 floor 0.0     # min:               0 @ t1 max:               1 @ t2
# No GLOBAL VARIABLES
NODAL VARIABLES relative 1.e-6 floor 0.0
	!u  # min:               0 @ t1,n1	max:               1 @ t2,n21
# No ELEMENT VARIABLES
# No NODESET VARIABLES
# No SIDESET VARIABLES
```

And the results when I ask exodiff to use it:
```
[#]> exodiff -f  simple.cmp gold/simple_diffusion_out.e simple_diffusion_out.e 
<... trimmed ...>
Nodal coordinates will be compared .. tol:    1e-06 (absolute), floor:        0
Time step values will be compared .. tol:    1e-06 (relative), floor:        0
No Global variables on either file.
Nodal variables will not be compared. <--- THIS IS STATED
No Element variables on either file.
No Element Attribute variables on either file.
No Nodeset variables on either file.
No Sideset variables on either file.
<... trimmed ...>

exodiff: Files are the same
```

As opposed to this when I don't:
```
<... trimmed ...>
Nodal coordinates will be compared .. tol:    1e-06 (absolute), floor:        0
Time step values will be compared .. tol:    1e-06 (relative), floor:        0
No Global variables on either file.
Nodal variables to be compared:
	u                                tol:    1e-06 (relative), floor:        0
No Element variables on either file.
No Element Attribute variables on either file.
No Nodeset variables on either file.
No Sideset variables on either file.
Sideset Distribution Factors will be compared .. tol:    1e-06 (relative), floor:        0
<... trimmed ...>
exodiff: Files are the same
```

permcody replied:
OK!

permcody commented on test/tests/test_harness/gold/b_csvdiff_comparison_out.csv:
```diff
@@ -0,0 +1,7 @@
+time,u_field
```
I'm not sure I like your override tests. If you only have one field to override, it's equivalent to a global override. Can you perhaps add at least one more column to these tests to verify that they actually behave differently than the global versions?

permcody reviewed (changes requested):
Looking really good! Your override tests need a bit more work. Also I'm a little confused on some of the behavior. `--only-compare-custom` is suppose to only compare the columns listed in `--custom-columns`, However, the doc string for custom-columns seems to indicate that that's the default behavior. So which is it? Does --custom-columns behave as an override for specific columns but also compare the remaining ones with the global, or does it only compare the custom? 

If we go with the way Exodiff works, listing columns in the comp file makes it so ONLY those values get compared. This is why the `-summary` option populates the skeleton file with a list of all columns so you can simply remove the ones you want to ignore. With that in mind, I don't even think you need the `--only-compare-custom` option at all! That should be the only mode for this tool. 

Part of what you are doing is making the tool we want. Don't worry about making it pass the Frankenstein tests that are checked in. Each of those tests just needs to be updated appropriately. This should simplify the logic a lot!

milljm commented:
I am sure there are some options we can remove entirely... for sure.
I'll do another pass over how all that argument logic is suppose to work. As I recall, I remember seeing that option (only-do-custom-column) and feeling un-easy about it.

permcody commented:
Ben just put in a PR to remove some of those. I might have one test that uses the other option, but it should also just die.

Commit:
Cleaned up unittests, removed custom-only argument

The overrides adjust only that field, while using defaults for the
rest. If one needs to only test a single field, a comparison file
will be required (rather than support --custom-only).

Tightened up the csvdiff unittests, so they are more believable
when they pass. Also made unittests a bit more straightforward as each
test has its own comparison/output file matching the name of the test
(removed the prereq/symbolic link combination).

permcody commented:
Yes, please remove it, if it breaks the testing, Ben and I can look at it and submit a separate PR or just add a commit to this one.

Commit:
Update requirement for missing field

Commit:
Forgot to remove the z_field

The z_field should be removed for a proper !z_field test.

permcody commented:
@milljm - are you still working on this PR?

milljm commented:
@permcody Nope... ready to be re-reviewed...

permcody commented:
@jason - will you rebase and repush? There are a lot of failures on this PR that I believe are all related to all of the issues we've had this week. I'm hoping a rebase will clean up the testing.
